File: Editorial/151_Reverse_Words_in_a_String/TypeScript/2_Reverse_the_Whole_String_and_Then_Reverse_Each_Word.ts
// Translated TypeScript solution
function reverseWords(s: string): string {
    // convert string to char array
    // and trim spaces at the same time
    let arr: string[] = trimSpaces(s);

    // reverse the whole string
    reverse(arr, 0, arr.length - 1);

    // reverse each word
    reverseEachWord(arr);

    return arr.join("");
}

function trimSpaces(s: string): string[] {
    let left = 0,
        right = s.length - 1;
    // remove leading spaces
    while (left <= right && s[left] === " ") ++left;
    // remove trailing spaces
    while (left <= right && s[right] === " ") --right;
    // reduce multiple spaces to single one
    let output: string[] = [];
    while (left <= right) {
        if (s[left] !== " ") output.push(s[left]);
        else if (output[output.length - 1] !== " ") output.push(s[left]);
        ++left;
    }
    return output;
}

function reverseEachWord(arr: string[]): void {
    let n = arr.length;
    let start = 0,
        end = 0;
    while (start < n) {
        // go to the end of the word
        while (end < n && arr[end] !== " ") ++end;
        // reverse the word
        reverse(arr, start, end - 1);
        // move to the next word
        start = end + 1;
        ++end;
    }
}

function reverse(arr: string[], start: number, end: number) {
    while (start < end) {
        const tmp = arr[start];
        arr[start++] = arr[end];
        arr[end--] = tmp;
    }
}

File: Editorial/151_Reverse_Words_in_a_String/CSharp/2_Reverse_the_Whole_String_and_Then_Reverse_Each_Word.cs
// Translated C# solution

public class Solution {
    public string ReverseWords(string s) {
        StringBuilder sb = TrimSpaces(s);
        // reverse the whole string
        Reverse(sb, 0, sb.Length - 1);
        // reverse each word
        ReverseEachWord(sb);
        return sb.ToString();
    }

    private StringBuilder TrimSpaces(string s) {
        int left = 0, right = s.Length - 1;
        // remove leading spaces
        while (left <= right && s[left] == ' ') ++left;
        // remove trailing spaces
        while (left <= right && s[right] == ' ') --right;
        // reduce multiple spaces to single one
        StringBuilder sb = new StringBuilder();
        while (left <= right) {
            if (s[left] != ' ') sb.Append(s[left]);
            else if (sb[sb.Length - 1] != ' ') sb.Append(s[left]);
            ++left;
        }

        return sb;
    }

    private void ReverseEachWord(StringBuilder sb) {
        int n = sb.Length;
        int start = 0, end = 0;
        while (start < n) {
            // go to the end of the word
            while (end < n && sb[end] != ' ') ++end;
            // reverse the word
            Reverse(sb, start, end - 1);
            // move to the next word
            start = end + 1;
            ++end;
        }
    }

    private void Reverse(StringBuilder sb, int left, int right) {
        while (left < right) {
            char tmp = sb[left];
            sb[left++] = sb[right];
            sb[right--] = tmp;
        }
    }
}
